File: lib/features/day/domain/day_extensions.dart
import 'package:live_vitalist/features/aliment/domain/aliment_extensions.dart';
import 'package:live_vitalist/features/aliment/data/aliment_bank.dart';
import 'package:live_vitalist/features/aliment/domain/aliment.dart';
import 'package:live_vitalist/features/day/domain/day.dart';
import 'package:live_vitalist/features/day/domain/meal.dart';

extension DaysGroupAnalysis on List<Day> {
  Day sum() {
    final merged = <String, Meal>{};

    for (final day in this) {
      for (final meal in day.meals) {
        merged.putIfAbsent(meal.name, () => Meal(name: meal.name));
        /* Creates a copy of the map aliment to solve the problem where
             the aliment was mutated when averaging multiple days. */
        merged[meal.name]!
            .aliments
            .addAll(meal.aliments.map((e) => e.copyWith()));
      }
    }

    return Day(meals: merged.values.toList());
  }

  Day average() {
    final Day sum = this.sum();

    final averagedMeals = sum.meals
        .map(
          (meal) => Meal(
            name: meal.name,
            aliments: meal.aliments
                .map((aliment) =>
                    aliment.copyWith(servingSize: aliment.servingSize / length))
                .toList(),
          ),
        )
        .toList();

    return Day(meals: averagedMeals);
  }
}

extension AlimentFlattening on Aliment {
  dynamic getKey() {
    return switch (this) {
      InstancedAliment a => a.alimentID,
      TemporaryAliment b => b, // The reference
    };
  }

  double basicServingSize(AlimentBankState bank) {
    return servingSize * readUnitSize(bank);
  }

  Aliment withBasicUnit(AlimentBankState bank) {
    return copyWith(
      unit: readDataRef(bank).unit,
      servingSize: basicServingSize(bank),
    );
  }

  Aliment addToAliment(Aliment current, AlimentBankState bank) {
    return current.copyWith(
      servingSize: current.servingSize + basicServingSize(bank),
    );
  }
}

extension DayAnalysis on Day {
  List<Aliment> get aliments => meals.expand((meal) => meal.aliments).toList();

  Map<String, double> readIntake(AlimentBankState bank) =>
      aliments.summedFields(bank);

  List<Aliment> totalAliments(AlimentBankState bank) {
    final Map<dynamic, Aliment> total = {}; // Working in basic unit
    for (final aliment in aliments) {
      total.update(
        aliment.getKey(),
        (current) => aliment.addToAliment(current, bank),
        ifAbsent: () => aliment.withBasicUnit(bank),
      );
    }
    return total.values.toList();
  }

  Map<Aliment, double> topIntakeAliments(
      String nutrient, AlimentBankState bank) {
    final ta = totalAliments(bank);
    final Map<Aliment, double> result = Map.fromEntries(ta.map(
        (aliment) => MapEntry(aliment, aliment.readField(nutrient, bank, 1.0))))
      ..removeWhere((_, value) => value == 0.0);

    return Map.fromEntries(
        result.entries.toList()..sort((a, b) => b.value.compareTo(a.value)));
  }
}

File: server/src/features/day/domain/DayExtensions.ts
import { Aliment, InstancedAliment } from "../../aliment/domain/Aliment";
import { AlimentBankState } from "../../aliment/domain/AlimentBankState";
import {
    readDataRef,
    readField,
    readUnitSize,
    summedFields,
} from "../../aliment/domain/AlimentExtensions";
import { Day } from "./Day";
import { Meal } from "./Meal";

export function sumDays(days: Day[]): Day {
    const merged: Record<string, Meal> = {};

    for (const day of days) {
        for (const meal of day.meals) {
            merged[meal.name] = {
                name: meal.name,
                aliments: [
                    ...merged[meal.name]?.aliments ?? [],
                    ...meal.aliments ?? [],
                ]
            };
        }
    }

    return { meals: Object.values(merged) };
}

export function averageDays(days: Day[]): Day {
    const sum = sumDays(days);

    const averagedMeals = sum.meals.map<Meal>(meal => ({
        name: meal.name,
        aliments: meal.aliments.map<Aliment>(aliment => ({
            ...aliment,
            servingSize: aliment.servingSize / days.length,
        })),
    }))

    return { meals: averagedMeals };
}

export function getAliments(day: Day): Aliment[] {
    return day.meals.flatMap(meal => meal.aliments);
}

export function getKey(aliment: Aliment): unknown {
    if ("alimentID" in aliment) {
        return (aliment as InstancedAliment).alimentID;
    }
    return aliment; // The reference
}

export function basicServingSize(
    aliment: Aliment, 
    bank: AlimentBankState
): number {
    return aliment.servingSize * readUnitSize(aliment, bank);
}

export function withBasicUnit(
    aliment: Aliment, 
    bank: AlimentBankState
): Aliment {
    return {
        ...aliment,
        unit: readDataRef(aliment, bank).unit,
        servingSize: basicServingSize(aliment, bank),
    };
}

export function addToAliment(
    aliment: Aliment, 
    current: Aliment, 
    bank: AlimentBankState
): Aliment {
    return {
        ...current,
        servingSize: current.servingSize + basicServingSize(aliment, bank),
    };
}

export function readDayIntake(
    day: Day, 
    bank: AlimentBankState
): Record<string, number> {
    const aliments = getAliments(day);
    return summedFields(aliments, bank);
}

export function totalAliments(
    day: Day, 
    bank: AlimentBankState
): Aliment[] {
    const total = new Map<unknown, Aliment>(); // Working in basic unit

    for (const aliment of getAliments(day)) {
        const key = getKey(aliment);
        const current = total.get(key);
        total.set(
            key,
            current === undefined
                ? withBasicUnit(aliment, bank)
                : addToAliment(aliment, current, bank),
        );
    }

    return [...total.values()];
}

export function topIntakeAliments(
    day: Day,
    nutrient: string,
    bank: AlimentBankState,
): Map<Aliment, number> {
    const entries = totalAliments(day, bank)
        .map<[Aliment, number]>(aliment => [aliment, readField(aliment, nutrient, bank, 1.0)])
        .filter(([, value]) => value !== 0.0)
        .sort((a, b) => b[1] - a[1]);

    return new Map(entries);
}
